FIREFLY-982: Add the capability to a create persistent resource from a table request

loitly opened the pull request:
https://jira.ipac.caltech.edu/browse/FIREFLY-982

This is taken from ticket under implementation notes.
```
Resources are persistent data taken from a unique Table Request. Uniqueness is based on getUniqueID.
Once a resource is created, it will remain unchanged until deleted.
firefly.util.table.makeResourceRequest is created to help work with resources.
See "Table as a Resource" in tests-table.html for sample usage.

Resource comes with basic security features. This info is saved as a .acl file next to the database files.

scope: This control the visibility of a resource. It can be 'global', 'user', or 'protected'.
    global is public. Anyone can access it. If this resource is created with a secret, then that secret is needed to delete it.
    user is visible to only the user who created it. User is based on a 'usrkey' cookie set on the browser.
    protected resource is one created with a secret token. It can only be accessed using the same secret token.
Any resource created without scope is considered global.
```

See test for an example of how it's used:
https://fireflydev.ipac.caltech.edu/firefly-982-table-as-resource/firefly/test/tests-table.html?test=Table+as+a+Resource

robyww commented on src/firefly/js/tables/TableRequestUtil.js:
```diff
@@ -160,6 +165,45 @@ export function makeVOCatalogRequest(title, params={}, options={}) {
+ * @func  makeResourceRequest
+ * @memberof firefly.util.table
+ */
+export function makeResourceRequest({resource={}, title, options={}, action='query'}) {
```
I think it function is too overloaded. It makes it API confusing. I think you should break it up it 

1. `makeResourceRequest` for query, which is consistent with how we do things now.
2. `createResource`
3. `deleteResource`

if you could come up with a good name then `createResource` and `deleteResource` could be combined (`putResource`?)

I had to really study your example to understand what was happning.

loitly replied:
That's what I had before.  `makeResourceRequest `,  and `createResource`.  
Then, I added the `delete` feature.  In order to not expose the `action` param, I need to add a third function, `deleteResource`.  That's when I decided to combine hem all and expose the `action` param.

In the simplest case, you can do this.  (original request)
```
const aResource = {request: makeFileRequest(null, 'https://web.ipac.caltech.edu/staff/roby/demo/WiseDemoTable.tbl')};
const myTbl =   makeResourceRequest({resource: aResource, title: 'My Table'});
firefly.showTable('actual', aResource);
```
I have no problem separate them into 3 functions.  After seeing my explanation, do you still think I should separate them?

robyww replied:
I talk to @loitly, he will break it into three calls, after that merge.

robyww commented on src/firefly/java/edu/caltech/ipac/firefly/server/query/ResourceProcessor.java:
```diff
@@ -0,0 +1,176 @@
+            @ParamDoc(name = "secret",          desc = "secret token used to access 'protected' resources"),
+        })
+public class ResourceProcessor extends EmbeddedDbProcessor {
+    public static final String PROC_ID = "ResourceProcessor";
```
I like how this just functions as a wrapper around other request.

robyww reviewed (changes requested):
Overall it looks great. Code is clean and I like the fact you included ACL features.

_Changes_: I am most concern the API naming. Breaking up `makeResourceRequest` into two or three functions will make it clearer.

Some smaller comments.
- Using different tables URLs in the examples - I included a list.
- I added some general comments in `ResourceProcessor.java` about things I have become more aware of in all my cleaning up recently.
- I don't see how you know to keep this DB permanently. How is that done?

loitly commented on src/firefly/java/edu/caltech/ipac/firefly/server/query/ResourceProcessor.java:
```diff
@@ -0,0 +1,176 @@
+        try {
+            TableServerRequest sreq = QueryUtil.getSearchRequest((TableServerRequest) request);
+            SortedSet<Param> params = sreq.getSearchParams();
+            return StringUtils.toString(params, "|");
```
Not only that.  Many of the util functions are out dated.  I noticed `combine` a few days ago while looking at your PR.  The overloaded functions with '[]' wrapped is so niched case that it should not be in a separate functions.
On top of that, with java 8.. you can do it like this
```
jshell> Arrays.asList("a","b").stream().collect(Collectors.joining("|"));
$2 ==> "a|b"
```
The `toString` function do have an advantage of not having to check for null, which is nice.

loitly commented on src/firefly/html/test/tests-table.html:
```diff
@@ -331,6 +331,67 @@ <h4>Show how to inject table meta using API</h4>
+
+            const {makeFileRequest, makeResourceRequest} = firefly.util.table;
+
+            // define 3 resources
```
Yes, but I purposely reuse the same URL to demonstrate how one can create multiple resources from the same URL.  I can use different tables if you think that info is not needed.

ejoliet reviewed (approved):
I think i understood that the added function needed for API has been addressed.
Question i have is what is the difference between 'user' and 'protected'?
Regardless, is fine with me.

loitly commented:
> I think i understood that the added function needed for API has been addressed. Question i have is what is the difference between 'user' and 'protected'? Regardless, is fine with me.

`user`:  only the user who created the resource can access it.
`protected`: when created with a password, anyone with the password can access it.

loitly commented:
@ejoliet @robyww I added these after your feedbacks.  Take a look if you have a chance, otherwise I'll merge later in the afternoon.
- split `makeResourceRequest` into 3 function, adding `createResource` and `deleteResource`.
- add `ResourceProcessorTest`
- rewrite  tests in `tests-table.html` to make it easier to understand

robyww commented on src/firefly/html/test/tests-table.html:
```diff
@@ -331,6 +331,62 @@ <h4>Show how to inject table meta using API</h4>
+            firefly.showTable('actual', accessDeniedTbl);
+
+            firefly.deleteResourceTests = function () {
+                makeResourceRequest({resource: global, action: 'delete'});
```
I think you meant `deleteResource` here

robyww reviewed (approved):
Found a small thing in the example but otherwise looks good.
